### ISSUE TITLE
"arithmetic_overflow" and "unconditional_panic" lints do not fire on `const fn` in bin crate that is only const-used

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#81265]
I tried the following code:
```rust
const fn overflow() -> u32 {
    0 - 1
}

pub const X: u32 = overflow();

fn main() {}
```
On stable, the emits a lint warning about the arithmetic overflow. On nightly, it does not.

This probably is related to https://github.com/rust-lang/rust/pull/78407. However, the part I do not understand is that I can only reproduce the problem with a binary crate, not with a library crate. Cc @oli-obk

**Comment by oli-obk** [comment]
That is odd indeed. I think it is a bug. Apparently the library crate will build the `optimized_mir` for `overflow` even though `overflow` is not public.

Is it a problem that we do not emit these lints on `const fn` that are only used in constants?

**Comment by RalfJung** [comment]
> Is it a problem that we do not emit these lints on const fn that are only used in constants?

Yeah I think so. We should properly emit lints for all `fn`, no matter if they are `const` or not and no matter if they are used or not and no matter if this is a bin crate or a lib crate.

Emiting lints as part of an optimizations makes this hard. To me that just shows it's a mistake to emit lints as part of an optimization.^^

**Comment by oli-obk** [comment]
Note that there are a bunch of lints that we do not emit in dead code. I feel like this is a similar situation. Either the lint is triggered on dead code in the const fn, or the interpretation of the const fn as part of the interpretation of a const item will report the same error anyway.

**Comment by RalfJung** [comment]
> Note that there are a bunch of lints that we do not emit in dead code.

If this is a dead code argument, then I assume we certainly *should* still emit the lint in a lib crate if the fn is `pub`?

> or the interpretation of the const fn as part of the interpretation of a const item will report the same error anyway

No error will be reported in release mode, since overflow checks are disabled. So the fact that there is an overflow is entirely lost when no lint is emitted.

**Comment by oli-obk** [comment]
>  I assume we certainly should still emit the lint in a lib crate if the fn is pub?

Yes, and we would, because libraries generate `optimized_mir` for all their pub functions

> No error will be reported in release mode, since overflow checks are disabled. So the fact that there is an overflow is entirely lost when no lint is emitted.

ugh, right.

So... if we want these lints unconditionally, we can run const prop just in analysis mode in the place where we also run borrowck and similar guaranteed MIR passes. in the optimization pipeline we'd then run const prop in mutating mode and possibly even not report *any* diagnostics from there.

**Comment by apiraino** [comment]
Assigning `P-medium` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2381265.20.22arithmetic_overflow.22.20and.20.22unconditional_panic.22.20lint.E2.80.A6) and removing `I-prioritize`.

**Comment by pietroalbini** [comment]
I'm a bit confused about this being labeled a regression, it seems to correctly emit an error on stable, beta and nightly.

**Comment by RalfJung** [comment]
@pietroalbini try doing a release build. Then it does not error on nightly, but AFAIK it used to error in the past (stable at the time the report was made). I will edit the OP to mention the "release" bit.

**Comment by pietroalbini** [comment]
Ugh. Then this does not only affect nightly, this affects *at least* stable 1.51.0.

**Comment by pietroalbini** [comment]
Yeah the PR you mention could've caused it landed in 1.51.0. Updated the labels accordingly.
